redis7/redis.go

```diff
@@ -0,0 +1,371 @@
+		opt(&cfg)
+	}
+
+	result := &store{
+		prefix: prefix,
+		mux:    &sync.RWMutex{},
+		cfg:    cfg,
+	}
+
+	result.log = cfg.Log
+	if result.log == nil {
+		result.log = logrus.StandardLogger()
+	}
+
+	client := redis.NewClient(clientOpts)
+
+	result.log.Debugf("Checking connection to Redis database (address=%s)...", client.Options().Addr)
+	_, err := client.Ping(context.TODO()).Result()
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to Redis database: %w", err)
+	}
+	result.log.Debug("Connection check successful")
+
+	result.client = client
+	result.rs = redsync.New(goredis.NewPool(client))
+
+	return result, nil
+}
+
+type store struct {
+	client *redis.Client
+	rs     *redsync.Redsync
+	log    *logrus.Logger
+	mux    *sync.RWMutex
+	// prefix contains a string that is prepended to each key, to simulate separate databases.
+	// Redis doesn't supported named databases but only preconfigured, (numerically) indexed databases
+	// which isn't very practical.
+	prefix string
+	cfg    stoabs.Config
+}
+
+func (s *store) Close(ctx context.Context) error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	s.log.Debug("Closing Redis store")
+	if s.client == nil {
+		// already closed
+		return nil
+	}
+	err := util.CallWithTimeout(ctx, s.client.Close, func() {
+		s.log.Error("Closing of Redis client timed out")
+	})
+	s.client = nil
+	return err
+}
+
+func (s *store) Write(fn func(stoabs.WriteTx) error, opts ...stoabs.TxOption) error {
+	if err := s.checkOpen(); err != nil {
+		return err
+	}
+
+	return s.doTX(func(writer redis.Pipeliner) error {
+		return fn(&tx{writer: writer, reader: s.client, store: s})
+	}, opts)
+}
+
+func (s *store) Read(fn func(stoabs.ReadTx) error) error {
+	if err := s.checkOpen(); err != nil {
+		return err
+	}
+
+	return fn(&tx{reader: s.client, store: s})
+}
+
+func (s *store) WriteShelf(shelfName string, fn func(stoabs.Writer) error) error {
+	if err := s.checkOpen(); err != nil {
+		return err
+	}
+	return s.doTX(func(tx redis.Pipeliner) error {
+		return fn(s.getShelf(shelfName, tx, s.client))
+	}, nil)
+}
+
+func (s *store) ReadShelf(shelfName string, fn func(stoabs.Reader) error) error {
+	if err := s.checkOpen(); err != nil {
+		return err
+	}
+	return fn(s.getShelf(shelfName, nil, s.client))
+}
+
+func (s *store) getShelf(shelfName string, writer redis.Cmdable, reader redis.Cmdable) *shelf {
+	return &shelf{
+		name:   shelfName,
+		prefix: s.prefix,
+		writer: writer,
+		reader: reader,
+		store:  s,
+	}
+}
+
+func (s *store) doTX(fn func(tx redis.Pipeliner) error, optsSlice []stoabs.TxOption) error {
+	if err := s.checkOpen(); err != nil {
+		return err
+	}
+
+	opts := stoabs.TxOptions(optsSlice)
+
+	// Obtain transaction-level write lock, if requested
+	var txMutex *redsync.Mutex
+	if opts.RequestsWriteLock() {
+		lockName := "lock_" + s.prefix
+		s.log.Tracef("Acquiring Redis distributed lock (name=%s)", lockName)
+		txMutex = s.rs.NewMutex(lockName)
+		err := txMutex.Lock()
+		if err != nil {
```

Comment:
What is the behaviour on locking. Does the 2nd TX get an error or does it wait for a sec and then gets an error?

Reply:
Second one retries (by default 32 times) after a set time (by default a random period between 50ms and 250ms). The initial lock will expire after a set time (by default 8 seconds). Now I think of it, we might want to rollback the transaction when the lock expired at the moment of commit (because another TX might've acquired it).

Reply:
Fixed.